Add API to allow transparent backgrounds when creating a PDF

rr-codes opened the pull request:
#### 6128bfc2feebff15fdde6e8823e0531da36005a8
<pre>
Add API to allow transparent backgrounds when creating a PDF
<a href="https://bugs.webkit.org/show_bug.cgi?id=250340">https://bugs.webkit.org/show_bug.cgi?id=250340</a>
rdar://76982433

Reviewed by Simon Fraser.

Adds an API to allow PDFs to have a transparent background when being created,
instead of always forcing a white background.

* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayers):
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/UIProcess/API/Cocoa/WKPDFConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKPDFConfiguration.mm:
(-[WKPDFConfiguration init]):
(-[WKPDFConfiguration copyWithZone:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawToPDF):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawToPDF):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PDFSnapshot.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/260328@main">https://commits.webkit.org/260328@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2667ce946667c47316c39a38e3c7914ec5e80598

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107873 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117011 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111762 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18323 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8244 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100053 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13827 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97042 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41530 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28680 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9860 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30028 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6924 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16012 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49616 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7143 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12137 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1464,7 +1464,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>
     IPC::Connection::AsyncReplyID drawRectToImage(WebFrameProxy*, const PrintInfo&, const WebCore::IntRect&, const WebCore::IntSize&, CompletionHandler<void(const WebKit::ShareableBitmapHandle&)>&&);
     IPC::Connection::AsyncReplyID drawPagesToPDF(WebFrameProxy*, const PrintInfo&, uint32_t first, uint32_t count, CompletionHandler<void(API::Data*)>&&);
-    void drawToPDF(WebCore::FrameIdentifier, const std::optional<WebCore::FloatRect>&, CompletionHandler<void(RefPtr<WebCore::SharedBuffer>&&)>&&);
+    void drawToPDF(WebCore::FrameIdentifier, const std::optional<WebCore::FloatRect>&, bool,  CompletionHandler<void(RefPtr<WebCore::SharedBuffer>&&)>&&);
```
name the bool.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1084,7 +1084,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 #endif
 
-    void drawToPDF(WebCore::FrameIdentifier, const std::optional<WebCore::FloatRect>&, CompletionHandler<void(RefPtr<WebCore::SharedBuffer>&&)>&&);
+    void drawToPDF(WebCore::FrameIdentifier, const std::optional<WebCore::FloatRect>&, bool,  CompletionHandler<void(RefPtr<WebCore::SharedBuffer>&&)>&&);
```
name the bool

smfr commented on Source/WebCore/rendering/BackgroundPainter.cpp:
```diff
@@ -134,6 +134,8 @@ void BackgroundPainter::paintFillLayers(const Color& color, const FillLayer& fil
 
     auto& context = m_paintInfo.context();
     auto baseBgColorUsage = BaseBackgroundColorUse;
+    if (m_paintInfo.paintBehavior & PaintBehavior::AllowTransparentBackground)
```
Which renderers is this consulted on? Is it only the root background that we want to make opaque? Why are we not sharing the code we use for making the view background transparent?

rr-codes replied:
> Which renderers is this consulted on

what does this mean exactly?

> Is it only the root background that we want to make opaque?

The intention is that for any part of the PDF where the background would be forced to white, to just leave it as transparent if it was originally transparent.

> Why are we not sharing the code we use for making the view background transparent?

I didn't know such code existed. Do you know where it is?

smfr replied:
Look around `FrameView::isTransparent()` and `RenderView::shouldPaintBaseBackground()`

rr-codes replied:
@smfr If I change those, wouldn't the web page itself change to become transparent? (instead of only having the transparency-ness apply to the generated PDF)

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5786,10 +5786,18 @@ void WebPage::drawToPDF(FrameIdentifier frameID, const std::optional<FloatRect>&
 
+    auto originalColor = frameView.baseBackgroundColor();
+    if (allowTransparentBackground)
+        frameView.updateBackgroundRecursively(Color::transparentBlack);
```
I'm not sure if it makes sense to do this on all subframes. If a subframe has a background color, it might need to print to avoid white-on-white text.

rr-codes replied:
ah ok. Then to make it only apply to the root background color, would I just use `setBaseBackgroundColor` instead of `updateBackgroundRecursively`?

rr-codes committed:
Add API to allow transparent backgrounds when creating a PDF
https://bugs.webkit.org/show_bug.cgi?id=250340
rdar://76982433

Reviewed by Simon Fraser.

Adds an API to allow PDFs to have a transparent background when being created,
instead of always forcing a white background.

* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayers):
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebKit/UIProcess/API/Cocoa/WKPDFConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKPDFConfiguration.mm:
(-[WKPDFConfiguration init]):
(-[WKPDFConfiguration copyWithZone:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawToPDF):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawToPDF):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PDFSnapshot.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/260328@main
